Patch 1 (#5902)
cpinitiative/usaco-guide: solutions/silver/cf-1991E.mdx

If the graph is bipartite, it can be split up into two groups, such that there are no edges connecting nodes within groups. Bob can assign nodes in group 1 with color 1, and nodes in group 2 with color 2. Once one group is fully complete, Bob can use either color 3 or the original color on the unfinished group, which ensures that nodes in different groups will always have different colors. Therefore, if the graph is bipartite, Bob will always win.

## Implementation

**Time Complexity:** $\mathcal{O}(N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
void solve() {
	int n, m;
	cin >> n >> m;
	vector<vector<int>> adj(n);

	for (int i = 0; i < m; i++) {
		int u, v;
		cin >> u >> v;
		adj[--u].push_back(--v);
		adj[v].push_back(u);
	}

	vector<int> c(n, -1);
	c[0] = 0;

	bool notBipartite = false;
	auto dfs = [&](auto &&self, int nd) -> void {
		for (int x : adj[nd]) {
			if (c[x] == -1) {
				c[x] = !c[nd];
				self(self, x);
			} else if (c[x] == c[nd]) {
				notBipartite = true;
				return;
			}
		}
	};

	dfs(dfs, 0);

> 	if (notBipartite) {

Reviewer: add comments

---
id: cf-1991E
source: CF
title: Coloring Game
author: Arnav Gokhale
---

## Explanation

Firstly, determine if the graph is bipartite.

> If the graph is not bipartite, the Alice can choose the same two colors everytime. Since the graph is not bipartite, it is impossible for Bob to color the graph with only two colors such that no connected nodes have the same color. Thus, if the graph is not bipartite then Alice will always win.

Reviewer: ```suggestion
If the graph is not bipartite, Alice can choose the same two colors every time. Since the graph is not bipartite, it is impossible for Bob to color the graph with only two colors such that no connected nodes have the same color. Thus, Alice will always win.
```

> If the graph is bipartite, it can be split up into two groups, such that there are no edges connecting nodes within groups. Bob can then choose to color nodes in group 1 with color 1 and nodes in group 2 with color 2. Once one group is fully complete, Bob can use either color 3 or the original color on the unfinished group, which ensures that nodes in different groups will always have different colors. Therefore, if the graph is bipartite then Bob will always win.

Reviewer: ```suggestion
If the graph is bipartite, it can be split up into two groups, such that there are no edges connecting nodes within groups. Bob can assign nodes in group 1 with color 1, and nodes in group 2 with color 2. Once one group is fully complete, Bob can use either color 3 or the original color on the unfinished group, which ensures that nodes in different groups will always have different colors. Therefore, if the graph is bipartite, Bob will always win.
```

## Implementation

**Time Complexity:** $\mathcal{O}(N)$

<LanguageSection>

<CPPSection>
```
#include <bits/stdc++.h>
using namespace std;
void solve() {

>     int n, m; cin >> n >> m;

Reviewer: ```suggestion
using namespace std;

void solve() {
    int n, m; 
    cin >> n >> m;
```

>     vector<vector<int>> a(n);

Reviewer: `adj` or `g` is better

    for (int i = 0; i < m; i++) {
        int u, v; cin >> u >> v;
        a[--u].push_back(--v);
        a[v].push_back(u);
    }

    vector<int> cl(n);

>     vector<int> tws, ows;

Reviewer: why not like, `colors`, `ones`, and `twos`

current var names don't really make sense lol

    bool f = false;
    auto dfs = [&] (auto&& self, int nd, int c) -> void {
        cl[nd] = c;
        if (c == 1) ows.push_back(nd);
        else tws.push_back(nd);

        for (int x : a[nd]) {
            if (cl[x] == 0) self(self, x, 3 - c);
            else if (cl[x] == c) {
                f = true; return;
            }
        }
    };

    dfs(dfs, 0, 1);

    if (f) {
        int gb;
        cout << "Alice" << endl;
        for (int i = 0; i < n; i++) {
            cout << 1 << " " << 2 << endl;
            cin >> gb >> gb;
        }
    } else {
        cout << "Bob" << endl;
        int c1, c2, pt1 = 0, pt2 = 0;

        for (int i = 0; i < n; i++) {
            cin >> c1 >> c2;
            if (c1 == 1 and c2 == 2 or c2 == 1 and c1 == 2) {
                if (pt1 != ows.size()) {

>                     cout << ows[pt1] + 1 << " " << 1 << endl; pt1++;

Reviewer: can you space these out
